clang/runtime/dpct-rt/include/dpct/codepin/detail/dispatcher.hpp

```diff
@@ -0,0 +1,232 @@
+//==---- dispatcher.hpp -----------------------------*-C++-*-------------==//
+//
+// Copyright (C) Intel Corporation
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===--------------------------------------------------------------------===//
+#ifndef __DPCT_CODEPIN_DISPATCHER_HPP__
+#define __DPCT_CODEPIN_DISPATCHER_HPP__
+
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <numeric>
+#include <sstream>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#ifdef __NVCC__
+#include <cuda_runtime.h>
+#else
+#include <dpct/dpct.hpp>
+#include <sycl/sycl.hpp>
+#endif
+
+namespace dpct {
+namespace experimental {
+namespace detail {
+
+inline static std::map<std::string, int> api_index;
+inline static std::vector<std::string> dump_json;
+inline static std::string dump_file = "dump_log.json";
+
+class Logger {
+public:
+  Logger(const std::string &dumpFile)
+      : dump_file(dumpFile), ipf(dumpFile, std::ios::in) {
+    if (ipf.is_open()) {
+      std::getline(ipf, data);
+      ipf.close();
+    }
+  }
+
+  ~Logger() {
+    opf.open(dump_file);
+    std::string json = std::accumulate(
+        dump_json.begin(), dump_json.end(), std::string("{"),
+        [](std::string acc, std::string val) { return acc + val + ','; });
+    if (!json.empty()) {
+      json.pop_back();
+    }
+    json += "}\n";
+    opf << json;
+    if (!opf.is_open()) {
+      opf.close();
+    }
+  }
+  const std::string &get_data() { return data; }
+
+private:
+  std::string dump_file;
+  std::ifstream ipf;
+  std::ofstream opf;
+  std::string data;
+};
+
+static Logger log(dump_file);
+
+inline std::map<void *, uint32_t> &get_ptr_size_map() {
+  static std::map<void *, uint32_t> ptr_size_map;
+  return ptr_size_map;
+}
+
+inline uint32_t get_pointer_size_in_bytes_from_map(void *ptr) {
+  const std::map<void *, uint32_t> &ptr_size_map = get_ptr_size_map();
+  const auto &it = ptr_size_map.find(ptr);
+  return (it != ptr_size_map.end()) ? it->second : 0;
+}
+
+#ifdef __NVCC__
+void dump(std::string &log, const int3 &value) {
+  log = log + '\t' + std::to_string(value.x);
```

Comment:
How about use the "
  std::ostringstream oss;
  oss << '\t' << value.x
      << '\t' << value.y
      << '\t' << value.z;
  
  log += oss.str();"